File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    static String[] nums;
    static HashSet<Integer> primeNums = new HashSet<>();
    static boolean[] vis = new boolean[9];
    
    public int solution(String numbers) {
        int answer = 0;
        
        nums = numbers.split("");
        
        
        dfs("", nums.length);
        
        
        
        return primeNums.size();
    }
    
    
    static void dfs(String numbers, int n){
        if(isPrime(numbers)){
            primeNums.add((Integer)Integer.parseInt(numbers));
        }
        
        for(int i=0;i<n;i++){
            if(vis[i]) continue;
            vis[i] = true;
            
            dfs(numbers+nums[i], n);
            
            vis[i] = false;
        }
    }
    
    static boolean isPrime(String numbers){
        if(numbers.equals("")) return false;
        int num = Integer.parseInt(numbers);
        if(num<=1) return false;
        if(num==2) return true;
        
        int sqrtNum = (int)Math.sqrt(num);
        
        for(int i=2;i<=sqrtNum;i++){
            if(num%i==0) return false;
        }
        
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.kt
import java.util.*
import kotlin.math.*

val primes = HashSet<Int>()
val vis = Array(8){false}

class Solution {
    fun solution(numbers: String): Int {
        var answer = 0
        
        val nums = numbers.split("").filter{it.isNotBlank()}.map{it.toInt()}
        
        dfs(nums, 0, nums.size)
        
        answer = primes.size
        
        return answer
    }
}

fun dfs(nums:List<Int>, number:Int, n:Int){
    if(isPrime(number)){
        primes.add(number)
    }
    
    for(i in 0 until n){
        if(vis[i]) continue
        vis[i] = true
        val num = nums[i]
        dfs(nums, number*10+num, n)
        vis[i] = false
    }
}

fun isPrime(num:Int):Boolean{
    if(num<=1) return false
    if(num==2) return true
    val end = sqrt(num.toDouble()).toInt()
    for(i in 2..end){
        if(num%i==0) return false
    }
    return true
}